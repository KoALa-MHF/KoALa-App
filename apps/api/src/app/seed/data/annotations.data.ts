import { Annotation } from '../../annotations/entities/annotation.entity';

export const AnnotationsData: Partial<Annotation>[] = [
  {
    id: 1,
    markerId: 1,
    userSessionId: 1,
    start: 200,
  },
];
