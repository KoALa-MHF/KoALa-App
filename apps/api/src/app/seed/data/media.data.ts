import { Media } from '../../media/entities/media.entity';

export const MediaData = [
  {
    id: 1,
    name: 'test.mp3',
    mimeType: 'audio/mpeg',
  },
] as [Media];
