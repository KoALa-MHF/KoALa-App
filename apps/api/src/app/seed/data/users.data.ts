export const UsersData = {
  sessionOwner1: {
    id: 1,
    displayName: 'Session Owner 1',
    email: 'session-owner-1@koala-app.de',
    samlId: '123456',
  },
  sessionOwner2: {
    id: 2,
    displayName: 'Session Owner 2',
    email: 'session-owner-2@koala-app.de',
    samlId: '654321',
  },
  sessionParticipant1: {
    id: 3,
    displayName: 'Session Participant 1',
    email: 'session-participant-1@koala-app.de',
  },
  sessionParticipant2: {
    id: 4,
    displayName: 'Session Participant 2',
    email: 'session-participant-2@koala-app.de',
  },
};
