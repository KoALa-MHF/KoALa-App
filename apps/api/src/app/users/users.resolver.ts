import { Resolver, Mutation, Args, Int, Query } from '@nestjs/graphql';
import { UsersService } from './users.service';
import { User } from './entities/user.entity';
import { UpdateUserInput } from './dto/update-user.input';
import { UseGuards } from '@nestjs/common';
import { AuthGuard } from '../core/guards/auth.guard';
import { CurrentUser } from '../core/decorators/user.decorator';

@Resolver(() => User)
export class UsersResolver {
  constructor(private readonly usersService: UsersService) {}

  @UseGuards(AuthGuard)
  @Query(() => User, { name: 'me' })
  me(@CurrentUser() user: User) {
    return this.usersService.findOne(user.id);
  }

  @UseGuards(AuthGuard)
  @Mutation(() => User)
  updateUser(
    @CurrentUser() user: User,
    @Args('updateUserInput')
    updateUserInput: UpdateUserInput
  ) {
    return this.usersService.update(id, updateUserInput);
  }
}
