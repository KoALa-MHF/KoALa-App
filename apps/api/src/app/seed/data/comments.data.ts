import { UsersData } from './users.data';

export const CommentsData = [
  {
    id: 1,
    owner: UsersData.sessionParticipant1,
    text: 'This is a comment from the Participant',
    annotationId: 1,
  },
  {
    id: 2,
    owner: UsersData.sessionOwner1,
    text: 'This is a comment from the Owner',
    annotationId: 1,
  },
];
