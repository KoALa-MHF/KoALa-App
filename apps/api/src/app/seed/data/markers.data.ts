import { MarkerType } from '../../markers/entities/marker.entity';
import { UsersData } from './users.data';

export const MarkersData = [
  {
    id: 1,
    type: MarkerType.EVENT,
    name: 'Marker 1',
    abbreviation: 'M1',
    description: 'Marker 1 Description',
    color: '#000000',
    icon: 'icon1',
    owner: UsersData.sessionOwner1,
    contentColor: '#FFFFFF',
  },
  {
    id: 2,
    type: MarkerType.RANGE,
    name: 'Marker 2',
    abbreviation: 'M2',
    description: 'Marker 2 Description',
    color: '#000000',
    icon: 'icon2',
    owner: UsersData.sessionOwner1,
    contentColor: '#FFFFFF',
  },
  {
    id: 3,
    type: MarkerType.SLIDER,
    name: 'Marker 3',
    abbreviation: 'M3',
    description: 'Marker 3 Description',
    color: '#000000',
    icon: 'icon3',
    rangeFrom: -5,
    rangeTo: 5,
    owner: UsersData.sessionOwner1,
    contentColor: '#FFFFFF',
  },
  {
    id: 4,
    type: MarkerType.EVENT,
    name: 'Marker 4',
    abbreviation: 'M4',
    description: 'Marker 4 Description',
    color: '#000000',
    icon: 'icon4',
    owner: UsersData.sessionOwner2,
    contentColor: '#FFFFFF',
  },
];
