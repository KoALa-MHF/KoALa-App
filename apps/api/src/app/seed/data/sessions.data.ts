import { SessionStatus } from '../../sessions/entities/session.entity';
import { UsersData } from './users.data';

export const SessionsData = [
  {
    id: 1,
    name: 'Session 1',
    owner: UsersData.sessionOwner1,
    status: SessionStatus.IN_PREPARATION,
  },
  {
    id: 2,
    name: 'Session 2',
    owner: UsersData.sessionOwner2,
    status: SessionStatus.CLOSED,
  },
  {
    id: 3,
    name: 'Session 3',
    owner: UsersData.sessionOwner1,
    status: SessionStatus.IN_PREPARATION,
    start: Date.now(),
    end: Date.now() + 1000000,
  },
  {
    id: 4,
    name: 'Session 4',
    owner: UsersData.sessionOwner1,
    status: SessionStatus.OPEN,
    start: Date.now() - 2000,
    end: Date.now() - 1000,
  },
  {
    id: 5,
    name: 'Session 5',
    owner: UsersData.sessionOwner1,
    status: SessionStatus.ARCHIVED,
  },
];
