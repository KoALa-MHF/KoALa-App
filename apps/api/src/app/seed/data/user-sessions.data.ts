import { UsersData } from './users.data';
import { SessionsData } from './sessions.data';

export const UserSessionsData = [
  {
    owner: UsersData.SESSION_PARTICIPANT_1,
    session: SessionsData[0],
  },
  {
    owner: UsersData.SESSION_PARTICIPANT_2,
    session: SessionsData[0],
  },
  {
    owner: UsersData.SESSION_OWNER_1,
    session: SessionsData[1],
  },
  {
    owner: UsersData.SESSION_PARTICIPANT_2,
    session: SessionsData[2],
  },
];
