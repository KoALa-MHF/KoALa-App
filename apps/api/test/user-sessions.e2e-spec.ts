import request, { LEGACY_WEBSOCKET_PROTOCOL, supertestWs } from 'supertest-graphql';
import gql from 'graphql-tag';
import { INestApplication } from '@nestjs/common';
import { setupTestApplication } from './mocks/test.util';
import { UsersData } from '../src/app/seed/data/users.data';

const QUERY_USER_SESSIONS = gql`
  query UserSessions {
    userSessions {
      id
      status
      note
      owner {
        id
        email
      }
      session {
        id
        name
      }
      annotations {
        id
        start
        end
      }
    }
  }
`;

const QUERY_USER_SESSION = gql`
  query UserSession($id: Int!) {
    userSession(id: $id) {
      id
      status
      note
      session {
        id
        name
      }
      annotations {
        id
        start
        end
        userSession {
          id
          status
        }
      }
    }
  }
`;

const UPDATE_USER_SESSION = gql`
  mutation UpdateUserSession($id: Int!, $updateUserSessionInput: UpdateUserSessionInput!) {
    updateUserSession(id: $id, updateUserSessionInput: $updateUserSessionInput) {
      id
      status
      note
      owner {
        id
        email
      }
      session {
        id
        name
        toolbars {
          id
        }
      }
      annotations {
        id
        marker {
          id
        }
      }
    }
  }
`;

const REMOVE_USER_SESSION = gql`
  mutation RemoveUserSession($id: Int!) {
    removeUserSession(id: $id) {
      id
    }
  }
`;

const QUERY_USER_SESSION_VARIABLES = {
  id: 5,
};

const UPDATE_USER_SESSION_VARIABLES = {
  id: 5,
  updateUserSessionInput: {
    note: 'Update Notes',
  },
};

const REMOVE_USER_SESSION_VARIABLES = {
  id: 5,
};

describe('User Sessions (e2e)', () => {
  let app: INestApplication;

  beforeEach(async () => {
    app = await setupTestApplication();
  });

  afterEach(async () => {
    await app.close();
  });

  describe('Query User Sessions', () => {
    it('Not authenticated user should get "Unauthorized" error', async () => {
      const { errors } = await request(app.getHttpServer()).query(QUERY_USER_SESSIONS);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Unauthorized');
    });

    it('Authenticated user should get list of all owned user sessions', async () => {
      const { data } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant2.id}`, { type: 'bearer' })
        .query(QUERY_USER_SESSIONS)
        .expectNoErrors();

      expect(data).toMatchSnapshot();
    });
  });

  describe('Query User Session', () => {
    it('Not authenticated user should get "Unauthorized" error', async () => {
      const { errors } = await request(app.getHttpServer())
        .query(QUERY_USER_SESSION)
        .variables(QUERY_USER_SESSION_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Unauthorized');
    });

    it('Authenticated user should get own user session', async () => {
      const { data } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant2.id}`, { type: 'bearer' })
        .query(QUERY_USER_SESSION)
        .variables(QUERY_USER_SESSION_VARIABLES)
        .expectNoErrors();

      expect(data).toMatchSnapshot();
    });

    it('Authenticated user must not get the user session of another user', async () => {
      const { errors } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant1.id}`, { type: 'bearer' })
        .query(QUERY_USER_SESSION)
        .variables(QUERY_USER_SESSION_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Forbidden');
    });
  });

  describe('Update User Session', () => {
    it('Not authenticated user should get "Unauthorized" error', async () => {
      const { errors } = await request(app.getHttpServer())
        .mutate(UPDATE_USER_SESSION)
        .variables(UPDATE_USER_SESSION_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Unauthorized');
    });

    it('Authenticated user can update own user session', async () => {
      const { data } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant2.id}`, { type: 'bearer' })
        .mutate(UPDATE_USER_SESSION)
        .variables(UPDATE_USER_SESSION_VARIABLES)
        .expectNoErrors();

      expect(data).toMatchSnapshot();
    });

    it('Authenticated user can only update the right fields', async () => {
      const { data } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant2.id}`, { type: 'bearer' })
        .mutate(UPDATE_USER_SESSION)
        .variables({
          ...UPDATE_USER_SESSION_VARIABLES,
          owner: {
            email: 'user_can_not_be_updated@koala-app.de',
          },
          sessionId: 10,
        })
        .expectNoErrors();

      expect(data).toMatchSnapshot();
    });

    it('Authenticated user cannot update user session of another user', async () => {
      const { errors } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant1.id}`, { type: 'bearer' })
        .mutate(UPDATE_USER_SESSION)
        .variables(UPDATE_USER_SESSION_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Forbidden');
    });
  });

  describe('Remove User Session', () => {
    it('Not authenticated user should get "Unauthorized" error', async () => {
      const { errors } = await request(app.getHttpServer())
        .mutate(REMOVE_USER_SESSION)
        .variables(REMOVE_USER_SESSION_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Unauthorized');
    });

    it('Authenticated user can remove own user session', async () => {
      const { data: userSessionsBeforeRemove } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant2.id}`, { type: 'bearer' })
        .query(QUERY_USER_SESSIONS)
        .expectNoErrors();

      expect(userSessionsBeforeRemove).toMatchSnapshot();

      const { data } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant2.id}`, { type: 'bearer' })
        .mutate(REMOVE_USER_SESSION)
        .variables(REMOVE_USER_SESSION_VARIABLES)
        .expectNoErrors();

      expect(data).toMatchSnapshot();

      const { data: userSessionsAfterRemove } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant2.id}`, { type: 'bearer' })
        .query(QUERY_USER_SESSIONS)
        .expectNoErrors();

      expect(userSessionsAfterRemove).toMatchSnapshot();
    });

    it('Authenticated user cannot remove a user session of another user', async () => {
      const { errors } = await request(app.getHttpServer())
        .auth(`${UsersData.sessionParticipant1.id}`, { type: 'bearer' })
        .mutate(REMOVE_USER_SESSION)
        .variables(REMOVE_USER_SESSION_VARIABLES);

      expect(errors).toHaveLength(1);
      expect(errors[0].message).toBe('Forbidden');
    });
  });
});
