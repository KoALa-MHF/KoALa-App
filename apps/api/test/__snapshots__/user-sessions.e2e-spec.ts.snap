// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Sessions (e2e) Create User Session Authenticated owner of session can create a user session 1`] = `
{
  "createUserSession": {
    "annotations": [],
    "id": 10,
    "note": "",
    "owner": {
      "email": "create-user-session-user@koala-app.de",
      "id": "5",
    },
    "session": {
      "id": "1",
      "name": "Session 1",
      "userSessions": [
        {
          "id": 1,
          "owner": {
            "email": "session-owner-1@koala-app.de",
            "id": "1",
          },
        },
        {
          "id": 6,
          "owner": {
            "email": "session-participant-1@koala-app.de",
            "id": "3",
          },
        },
        {
          "id": 7,
          "owner": {
            "email": "session-participant-2@koala-app.de",
            "id": "4",
          },
        },
        {
          "id": 10,
          "owner": {
            "email": "create-user-session-user@koala-app.de",
            "id": "5",
          },
        },
      ],
    },
    "status": "INITIAL",
  },
}
`;

exports[`User Sessions (e2e) Invite User Session Authenticated owner of session can invite users to a session 1`] = `
{
  "inviteUserSession": [
    {
      "id": 1,
      "session": {
        "id": "1",
      },
      "status": "INITIAL",
    },
    {
      "id": 2,
      "session": {
        "id": "2",
      },
      "status": "INITIAL",
    },
    {
      "id": 3,
      "session": {
        "id": "3",
      },
      "status": "INITIAL",
    },
    {
      "id": 7,
      "session": {
        "id": "1",
      },
      "status": "INITIAL",
    },
  ],
}
`;

exports[`User Sessions (e2e) Query User Session Authenticated user should get own user session 1`] = `
{
  "userSession": {
    "annotations": [],
    "id": 7,
    "note": "",
    "session": {
      "id": "1",
      "name": "Session 1",
    },
    "status": "INITIAL",
  },
}
`;

exports[`User Sessions (e2e) Query User Sessions Authenticated user should get list of all owned user sessions 1`] = `
{
  "userSessions": [
    {
      "annotations": [],
      "id": 7,
      "note": "",
      "owner": {
        "email": "session-participant-2@koala-app.de",
        "id": "4",
      },
      "session": {
        "id": "1",
        "name": "Session 1",
      },
      "status": "INITIAL",
    },
    {
      "annotations": [],
      "id": 9,
      "note": "",
      "owner": {
        "email": "session-participant-2@koala-app.de",
        "id": "4",
      },
      "session": {
        "id": "3",
        "name": "Session 3",
      },
      "status": "INITIAL",
    },
  ],
}
`;

exports[`User Sessions (e2e) Remove User Session Authenticated user can remove own user session 1`] = `
{
  "userSessions": [
    {
      "annotations": [],
      "id": 7,
      "note": "",
      "owner": {
        "email": "session-participant-2@koala-app.de",
        "id": "4",
      },
      "session": {
        "id": "1",
        "name": "Session 1",
      },
      "status": "INITIAL",
    },
    {
      "annotations": [],
      "id": 9,
      "note": "",
      "owner": {
        "email": "session-participant-2@koala-app.de",
        "id": "4",
      },
      "session": {
        "id": "3",
        "name": "Session 3",
      },
      "status": "INITIAL",
    },
  ],
}
`;

exports[`User Sessions (e2e) Remove User Session Authenticated user can remove own user session 2`] = `
{
  "removeUserSession": {
    "id": 7,
  },
}
`;

exports[`User Sessions (e2e) Remove User Session Authenticated user can remove own user session 3`] = `
{
  "userSessions": [
    {
      "annotations": [],
      "id": 9,
      "note": "",
      "owner": {
        "email": "session-participant-2@koala-app.de",
        "id": "4",
      },
      "session": {
        "id": "3",
        "name": "Session 3",
      },
      "status": "INITIAL",
    },
  ],
}
`;

exports[`User Sessions (e2e) Update User Session Authenticated user can only update the right fields 1`] = `
{
  "updateUserSession": {
    "annotations": [],
    "id": 7,
    "note": "Update Notes",
    "owner": {
      "email": "session-participant-2@koala-app.de",
      "id": "4",
    },
    "session": {
      "id": "1",
      "name": "Session 1",
      "toolbars": [
        {
          "id": "1",
        },
      ],
    },
    "status": "INITIAL",
  },
}
`;

exports[`User Sessions (e2e) Update User Session Authenticated user can update own user session 1`] = `
{
  "updateUserSession": {
    "annotations": [],
    "id": 7,
    "note": "Update Notes",
    "owner": {
      "email": "session-participant-2@koala-app.de",
      "id": "4",
    },
    "session": {
      "id": "1",
      "name": "Session 1",
      "toolbars": [
        {
          "id": "1",
        },
      ],
    },
    "status": "INITIAL",
  },
}
`;
